cudax/include/cuda/experimental/__memory_resource/async_memory_resource.h

```diff
@@ -0,0 +1,298 @@
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the specified device.
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails.
+  //! @returns The default memory pool of the specified device.
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool(const int __device_id)
+  {
+    ::cuda::experimental::mr::__device_supports_stream_ordered_allocations(__device_id);
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the async_memory_resource using the default \c cudaMemPool_t of the default device.
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails.
+  async_memory_resource()
+      : __pool_(__get_default_mem_pool(0))
+  {}
+
+  //! @brief Constructs a async_memory_resource using the default \c cudaMemPool_t of a given device.
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails.
+  async_memory_resource(::cuda::experimental::device_ref __device)
+      : __pool_(__get_default_mem_pool(__device.get()))
+  {}
+
+  async_memory_resource(int)                   = delete;
+  async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the async_memory_resource from a \c async_memory_pool by calling get().
+  //! @param __pool The \c async_memory_pool used to allocate memory.
+  explicit async_memory_resource(async_memory_pool& __pool) noexcept
+      : __pool_(__pool.get())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws std::invalid_argument In case of invalid alignment.
+  //! @throws cuda::cuda_error If an error code was return by the CUDA API call.
+  //! @returns Pointer to the newly allocated memory.
+  _CCCL_NODISCARD void* allocate(const size_t __bytes,
+                                 const size_t __alignment = _CUDA_VMR::default_cuda_malloc_alignment)
+  {
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_invalid_argument(
+        "Invalid alignment passed to "
+        "async_memory_resource::allocate_async.");
+    }
+
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocFromPoolAsync,
+      "async_memory_resource::allocate failed to allocate with cudaMallocFromPoolAsync",
+      &__ptr,
+      __bytes,
+      __pool_,
+      ::cudaStream_t{0});
```

Comment:
can you remind me why we're using the null stream here? i thought we were trying to get away from that.

Reply:
There was another comment thread where I mentioned we can't claim its a synchronous memory allocation function if we don't synchronize, even if we use the NULL stream.
I think we mentioned just throwing in the synchronous version, but I might have a different solution.

In the cudaMallocFromPoolAsync doc it says:
```
The specified memory pool may be from a device different than that of the specified hStream.
```
I think we could add a static global stream on device 0, that would serve as a backup stream for memory allocation/deallocation in cases where we don't have any dependencies. I think there shouldn't be any issues with false dependencies and with synchronization of that stream, as long as we only use it for quickly resolving memory operations. Any thoughts?

Reply:
I have discussed this with the rapids folks. 

The issue they found with using a unique stream to synchronize is that the user has no way of knowing which stream has been used.

This makes it effectively impossible to use that API and pass it back to an asynchronous interface. That might be a nice use case but it is still technically valid.

Knowing that the stream that we pass does not need to match the device the memory was originally allocated on, I believe that always using the default stream of the current device is the right way forward until we just completely rip the synchronous API out

Reply:
But wait, the implementation is not even synchronous as is, it is still making the asynchronous allocation, just on the NULL stream. If it's used in a non-blocking stream it's going to blow up.

Something needs to change, we either change the semantics of this API (to what?) or we change the implementation.

Reply:
Oh wait we are missing the stream synchronize